auth/gcloud/aio/auth/token.py

```diff
@@ -267,18 +326,158 @@ async def acquire_access_token(self, timeout: int = 10) -> None:
+        }
+        payload = json.dumps({
+            'includeEmail': True,
+            'audience': iap_client_id,
+        })
+        resp = await self.session.post(
+            GCLOUD_ENDPOINT_GENERATE_ID_TOKEN.format(
+                service_account=self.service_account),
+            data=payload, headers=headers, timeout=timeout)
+
         content = await resp.json()
+        return TokenResponse(value=content['token'])
+
+    async def _refresh_gce_metadata(self, timeout: int) -> TokenResponse:
+        """
+        Fetch IAP ID token from the GCE metadata servers.
+
+        https://cloud.google.com/docs/authentication/get-id-token#metadata-server
+        """
+        resp = await self.session.get(
+            GCE_ENDPOINT_ID_TOKEN.format(audience=self.app_uri),
+            headers=GCE_METADATA_HEADERS, timeout=timeout,
+            allow_redirects=True)
+        token = await resp.text()
+        return TokenResponse(value=token)
+
+    async def _refresh_service_account(
+        self, iap_client_id: str,
+        timeout: int,
+    ) -> TokenResponse:
+        now = int(time.time())
+        expiry = now + GCLOUD_TOKEN_DURATION
 
-        self.access_token = str(content['access_token'])
-        self.access_token_duration = int(content['expires_in'])
-        self.access_token_acquired_at = datetime.datetime.utcnow()
-        self.acquiring = None
+        assertion_payload = {
+            'iss': self.service_data['client_email'],
+            'aud': self.token_uri,
+            'exp': expiry,
+            'iat': now,
+            'sub': self.service_data['client_email'],
+            'target_audience': iap_client_id,
+        }
 
-    async def close(self) -> None:
-        await self.session.close()
+        assertion = jwt.encode(
+            assertion_payload,
+            self.service_data['private_key'],
+            algorithm='RS256',
+        )
 
-    async def __aenter__(self) -> 'Token':
-        return self
+        payload = urlencode({
+            'assertion': assertion,
+            'grant_type': 'urn:ietf:params:oauth:grant-type:jwt-bearer',
+        })
 
-    async def __aexit__(self, *args: Any) -> None:
-        await self.close()
+        resp = await self.session.post(self.token_uri, data=payload,
+                                       headers=REFRESH_HEADERS,
+                                       timeout=timeout)
+
+        content = await resp.json()
+        return TokenResponse(value=content['id_token'],
+                             expires_in=expiry - int(time.time()))
+
+    async def refresh(self, *, timeout: int) -> TokenResponse:
+        if self.token_type == Type.AUTHORIZED_USER:
+            iap_client_id = await self._get_iap_client_id(timeout=timeout)
+            resp = await self._refresh_authorized_user(
+                iap_client_id, timeout=timeout)
+        elif self.token_type == Type.GCE_METADATA:
+            resp = await self._refresh_gce_metadata(timeout=timeout)
```

Comment:
Huh, how interesting that the client ID isn't necessary in this flow!

Reply:
Agreed!